src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +23,78 @@
 class ExceptionSink(object):
   """A mutable singleton object representing where exceptions should be logged to."""
 
-  _destination = os.getcwd()
+  # TODO: see the bottom of this file where we call set_destination() and friends in order to
+  # properly setup global state.
+  # TODO: ???
+  _destination = None
+  # We need an exiter in order to know what to do after we log a fatal exception.
+  _exiter = None
+  # Where to log stacktraces to in a SIGUSR2 handler.
+  _trace_stream = None
+
+  _all_destinations = []
 
   def __new__(cls, *args, **kwargs):
     raise TypeError('Instances of {} are not allowed to be constructed!'
                     .format(cls.__name__))
 
+  class ExceptionSinkError(Exception): pass
+
+  # TODO (in response to the above): we can avoid a single entry point and make everything work with
+  # great ease if we can ensure all set_* methods are idempotent!
   @classmethod
   def set_destination(cls, dir_path):
-    cls._destination = dir_path
+    cls._destination = cls._check_or_create_new_destination(dir_path)
+    cls._all_destinations.append(cls._destination)
```

Comment:
After thinking about the use case of RemotePantsRunner a bit, I think I agree, because if we're going to want to dive into any exception logging like we do there we'll probably know where to look instead of wanting to dredge up all our possible log files (?).

However, after thinking, I'm thinking now mostly about the situation which is completely unhandled now, which is that if something happens to call `ExceptionSink.set_destination()` after RemotePantsRunner does, we immediately lose any logs we might have wanted for the thin client's error message.

The `_all_destinations` bit doesn't really address that except by way of the commented-out incomplete method body which would just dredge up all the logs (not useful for the same/opposite reason), but that situation might be something we want to special-case.